core/solver/multigrid.cpp

```diff
@@ -159,346 +160,338 @@ void handle_list(
+        // allocate the previous level
+        g_list.emplace_back(vec::create(exec, dim<2>{current_nrows, nrhs}));
+        e_list.emplace_back(vec::create(exec, dim<2>{current_nrows, nrhs}));
+        next_one_list.emplace_back(initialize<vec>({one<ValueType>()}, exec));
+    }
+    if (level + 1 == multigrid->get_mg_level_list().size()) {
+        // the last level allocate the g, e for coarsest solver
+        g_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
+        e_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
+        next_one_list.emplace_back(initialize<vec>({one<ValueType>()}, exec));
+    }
+    one_list.emplace_back(initialize<vec>({one<ValueType>()}, exec));
+    neg_one_list.emplace_back(initialize<vec>({-one<ValueType>()}, exec));
+}
+
+
+void MultigridState::run_cycle(multigrid::cycle cycle, size_type level,
+                               const std::shared_ptr<const LinOp>& matrix,
+                               const LinOp* b, LinOp* x, cycle_info info)
+{
+    if (level == multigrid->get_mg_level_list().size()) {
+        multigrid->get_coarsest_solver()->apply(b, x);
+        return;
+    }
+    auto mg_level = multigrid->get_mg_level_list().at(level);
+    run<gko::multigrid::EnableMultigridLevel, float, double,
+        std::complex<float>, std::complex<double>>(
+        mg_level, [&, this](auto mg_level) {
+            using value_type = typename std::decay_t<
+                gko::detail::pointee<decltype(mg_level)>>::value_type;
+            this->run_cycle<value_type>(cycle, level, matrix, b, x, info);
+        });
+}
+
+
+template <typename ValueType>
+void MultigridState::run_cycle(multigrid::cycle cycle, size_type level,
+                               const std::shared_ptr<const LinOp>& matrix,
+                               const LinOp* b, LinOp* x, cycle_info info)
+{
+    auto total_level = multigrid->get_mg_level_list().size();
+
+    auto r = r_list.at(level);
+    auto g = g_list.at(level);
+    auto e = e_list.at(level);
+    // get mg_level
+    auto mg_level = multigrid->get_mg_level_list().at(level);
+    // get the pre_smoother
+    auto pre_smoother = multigrid->get_pre_smoother_list().at(level);
+    // get the mid_smoother
+    std::shared_ptr<const LinOp> mid_smoother{nullptr};
+    auto mid_case = multigrid->get_parameters().mid_case;
+    if (mid_case == multigrid::mid_smooth_type::standalone) {
+        mid_smoother = multigrid->get_mid_smoother_list().at(level);
+    }
+    // get the post_smoother
+    auto post_smoother = multigrid->get_post_smoother_list().at(level);
+    auto one = one_list.at(level).get();
+    auto next_one = next_one_list.at(level).get();
+    auto neg_one = neg_one_list.at(level).get();
+    // origin or next or first
+    bool use_pre = (info & cycle_info::first_of_cycle) ||
+                   mid_case == multigrid::mid_smooth_type::both ||
+                   mid_case == multigrid::mid_smooth_type::pre_smoother;
```

Comment:
Do we check for the case if the user wants a pre-smoother (has used the `mid_smooth_type::pre_smoother` option)  but has not provided a pre_smoother list ?

Reply:
it will behave like pre_smoother, so it will not process smoother